flink-autoscaler/src/main/java/org/apache/flink/autoscaler/ScalingMetricEvaluator.java

```diff
@@ -296,8 +297,14 @@ protected static void computeProcessingRateThresholds(
             upperUtilization = 1.0;
             lowerUtilization = 0.0;
         } else {
-            upperUtilization = targetUtilization + utilizationBoundary;
-            lowerUtilization = targetUtilization - utilizationBoundary;
+            upperUtilization =
+                    conf.getOptional(TARGET_UTILIZATION_BOUNDARY)
+                            .map(boundary -> targetUtilization + boundary)
+                            .orElseGet(() -> conf.get(UTILIZATION_MAX));
+            lowerUtilization =
+                    conf.getOptional(TARGET_UTILIZATION_BOUNDARY)
+                            .map(boundary -> targetUtilization - boundary)
+                            .orElseGet(() -> conf.get(UTILIZATION_MIN));
         }
 
         double scaleUpThreshold =
```

Comment:
In `AutoScalerUtils.getTargetProcessingCapacity`, we adjust `targetUtilization` between 0.0 and 1.0, it's reasonable.

After introducing `UTILIZATION_MAX` and `UTILIZATION_MIN`, it's better to ensure:
-  1.0 >= UTILIZATION_MAX >= UTILIZATION_TARGET
- UTILIZATION_TARGET >= UTILIZATION_MIN >= 0.0

And it's needed to add some tests to check them.

Also, could we unify the logic in this method instead of `AutoScalerUtils.getTargetProcessingCapacity`?


Of course, another solution is when `UTILIZATION_MAX < UTILIZATION_TARGET` or `UTILIZATION_TARGET < UTILIZATION_MIN`, we could throw exception.

Reply:
Thanks for review.

The last commit was not fully checked and there were some errors in some places, which have been fixed.


> In `AutoScalerUtils.getTargetProcessingCapacity`, we adjust `targetUtilization` between 0.0 and 1.0, it's reasonable.
> 
> After introducing `UTILIZATION_MAX` and `UTILIZATION_MIN`, it's better to ensure:
> 
> * 1.0 >= UTILIZATION_MAX >= UTILIZATION_TARGET
> * UTILIZATION_TARGET >= UTILIZATION_MIN >= 0.0
> 
> And it's needed to add some tests to check them.
> 

Added `DefaultValidator#validateNumberOrder` to verify that multiple parameters need to follow the logic of size order, and added test cases. Also`validateNumber` can also limit the maximum value

> Also, could we unify the logic in this method instead of `AutoScalerUtils.getTargetProcessingCapacity`?

`AutoScalerUtils.getTargetProcessingCapacity` I think it is reasonable to keep them in this pr, Because it is used in many places

Reply:
`DefaultValidator` is a class of `flink-kubernetes-operator` module. It doesn't work for Autoscaler Standalone or other scenarios. Could we validate autoscaler related options inside of `flink-autoscaler` module? and call it in `JobAutoScalerImpl#scale`?

Reply:
> But how does Autoscaler standlone handle exceptions where these parameters are unreasonable.
by event report?
> 
> I'm not sure if it needs to be resolved via another PR?

It's acceptable that Autoscaler standlone check options in a separate PR.